periodicity: added support for modulo operation

akash9712 opened the pull request:
`periodicity(f, symbol)` uses decompogen to decompose a function, and
uses recursion until one of the function in `decompogen (f,symbol)`
has its periodicity implemented as a base case. As of this commit, `Mod`
function is one of the base cases, so periodicity of modulo function can be
evaluated as well.
```
Example:
In [1]: periodicity(tan(3*x%4),x)
Out[1]: 4/3

In [2]: periodicity(tan((x**2)%3),x)

In [3]: periodicity(sin(x)%3,x)
Out[3]: 2⋅π
```
<!-- Please give this pull request a descriptive title. Pull requests with descriptive titles are more likely to receive reviews. Describe what you changed! A title that only references an issue number is not descriptive. -->

<!-- If this pull request fixes an issue please indicate which issue by typing "Fixes #NNNN" below. -->

akash9712 committed:
periodicity: added support for modulo operation

periodicity(f, symbol) uses decompogen to decompose a function, and
uses recursion until one of the function in decompogen in (f,symbol)
has its periodicity implemented as a base case. As of this commit, Mod
function is one of the base cases, so periodicity of modulo function can be
evaluated as well.

Example:
In [1]: periodicity(tan(3*x%4),x)
Out[1]: 4/3

In [2]: periodicity(tan((x**2)%3),x)

In [3]: periodicity(sin(x)%3,x)
Out[3]: 2⋅π

smichr commented on sympy/calculus/util.py:
```diff
@@ -373,6 +375,16 @@ def periodicity(f, symbol, check=False):
         period = _periodicity(g.args, symbol)
 
+    elif isinstance(f, Mod):
+        a,n=f.args
```
space after comma
space around `==`
space around  `=` except in arg list
(see [pep8](https://www.python.org/dev/peps/pep-0008/))

smichr commented on sympy/calculus/util.py:
```diff
@@ -373,6 +375,16 @@ def periodicity(f, symbol, check=False):
+        a,n=f.args
+
+        if a==symbol:
+                period=n
```
indent only 4 here

smichr commented on sympy/calculus/tests/test_util.py:
```diff
@@ -85,6 +85,9 @@ def test_periodicity():
     assert periodicity(cos(sec(x) - csc(2*x)), x) == 2*pi
     assert periodicity(tan(sin(2*x)), x) == pi
     assert periodicity(2*tan(x)**2, x) == pi
+    assert periodicity(sin(x%4),x)==4
```
space after comma, around any of `%+-`, and around `==`

akash9712 committed:
Edited whitespaces

Added whitespaces in test_util.py and util.py

smichr commented on sympy/calculus/util.py:
```diff
@@ -373,6 +375,16 @@ def periodicity(f, symbol, check=False):
+        a, n = f.args
+
+        if a == symbol:
+            period  = n
```
The extra space to the left of the `==` can be moved to up above (in the test file) where there are no spaces ;-) (lines 88-90 and line 104.

akash9712 replied:
Oops. Edited accordingly. :)

akash9712 replied:
Oops. Edited accordingly. :)

akash9712 committed:
Edited whitespaces

removed whitespace from util.py, and added to test_util.py

akash9712 committed:
Merge branch 'master' of https://github.com/sympy/sympy into Mod_periodicity

smichr commented on sympy/calculus/util.py:
```diff
@@ -373,6 +375,16 @@ def periodicity(f, symbol, check=False):
+        elif isinstance(a, TrigonometricFunction):
+            period = periodicity(a, symbol)
+        elif degree(a, symbol) == 1: #check if 'f' is linear in 'symbol'
+            period = Abs(n/a.coeff(symbol))
```
this is not robust enough:

```
>>> (sqrt(2)*(x+1)-x).coeff(x)
-1 # instead of sqrt(2) - 1
>>> (sqrt(2)*(x+1)-x).diff(x)
-1 + sqrt(2)
```

I recommend that you use `Abs(n/a.diff(symbol))`

akash9712 committed:
Merge branch 'master' of https://github.com/sympy/sympy into Mod_periodicity

akash9712 committed:
Added conditions for Mod in periodicity

A condition is added to check if n in Mod(a,n) is independent of the
symbol for which the periodicity is to be calculated. Also,`a.coeff(symbol)`
is replaced with `a.diff(symbol)`. Tests were added.

smichr commented:
There are conflicts to resolve.

akash9712 committed:
Merge branch 'master' into Mod_periodicity

akash9712 commented:
@smichr The merge conflicts have been resolved. Please review.

smichr commented:
Thanks, it's in.
